File: kafkatest/go/consumer.go
package main

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/confluentinc/confluent-kafka-go/kafka"
)

func consumer(url string) {
	consumer, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": url,
		"group.id":          "rust",
		"auto.offset.reset": "earliest",
	})
	if err != nil {
		panic(err)
	}
	defer consumer.Close()

	err = consumer.SubscribeTopics([]string{"rust_topic"}, nil)
	if err != nil {
		log.Fatal(err)
	}

	for {
		msg, err := consumer.ReadMessage(-1)
		if err == nil {
			var ev Event
			if err := json.Unmarshal(msg.Value, &ev); err == nil {
				fmt.Printf("Received struct: %v\n", ev)
			} else {
				fmt.Println("Invalid JSON")
			}
		} else {
			fmt.Printf("Consumer error: %v (%v)\n", err, msg)
		}
	}
}

File: kafkatest/rust/src/consumer.rs
use rdkafka::{
    ClientConfig, Message,
    consumer::{BaseConsumer, CommitMode, Consumer},
};

use crate::types::Event;

pub async fn start() {
    let consumer = create();
    consume(consumer).await;
}

fn create() -> BaseConsumer {
    let url = std::env::var("KAFKA_BROKER").unwrap_or("localhost:9092".to_string());

    let mut config = ClientConfig::new();
    config
        .set("bootstrap.servers", &url)
        .set("group.id", "go")
        .set("enable.auto.commit", "false")
        .set("auto.offset.reset", "earliest");

    config.create().expect("Failed to create consumer")
}

async fn consume(consumer: BaseConsumer) {
    consumer.subscribe(&["go_topic"]).expect("Cannot subscribe");

    loop {
        match consumer.poll(None) {
            None => {
                continue;
            }
            Some(Err(e)) => println!("Kafka error: {:?}", e),
            Some(Ok(msg)) => {
                if let Some(payload) = msg.payload() {
                    let event: Result<Event, _> = serde_json::from_slice(payload);
                    match event {
                        Ok(ev) => println!("Received struct: {:?}", ev),
                        Err(_) => println!("Invalid JSON"),
                    }
                } else {
                    println!("No message payload or invalid UTF-8");
                }
                consumer.commit_message(&msg, CommitMode::Async).unwrap();
            }
        }
    }
}
